### DISCUSSION TITLE
Access to $event and $dispatch in event handlers added via x-spread

### DISCUSSION THREAD

**Discussion by allmarkedup** [alpinejs/alpine#617] [4. General]
Is it possible to access the native browser Event object (`$event`) within an event listener that has been added using the new x-spread syntax?

Using normal attribute binding I would pass this in (as for the $dispatch utility) like this:

```html
<button @click="handleClick($event, $dispatch)"></button>
```

But using x-spread like so I'm not sure how to access them. `this.$event` does not work and no arguments are passed into the event handler function. 

```html
<div x-data="Example()">
  <button x-spread="button"></button>
</div>
```
```js
function Example(){
  return {
    button: {
      ['@click'](){
         // how to access $event and $dispatch here?
      }
    }
  }
}
```

Any help would be much appreciated!

**Comment by SimoTod** [comment]
I don't think that syntax works with magic properties at the moment. You should simulate an inline handlers because it's where $event is defined. 
I would expect it to be
```
function Example(){
  return {
    button: {
      ['@click']: 'foo' // or 'foo($event)' 
    },
    foo($event) {
    } 
  }
}
```

**Reply by allmarkedup** [reply]
Ah perfect thank you @SimoTod - I hadn't thought of defining it like that but that makes perfect sense as the equivalent to using `@click="handleClick` directly on the element (and thus getting the event handler supplied as the first argument). Looks neater too in my opinion :-) 

Thanks again!

**Reply by northern-mindset** [reply]
Why do we use square brackets for the property name? According to the [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#computed_property_names), square brackets are used for computed property names, but we are not computing anything, we use string literals inside.

Actually, the following (from a real project) works just fine as well:

```
self: {
  'x-show.transition': 'open',
  '@archive:toggle-play.window'({ detail }) {
    this.fileId = detail.fileId;
    this.metadata = detail.metadata;
    this.open = true;
  },
}
```

**Reply by SimoTod** [reply]
Thanks for your feedback, you are absolutely correct and you can omit the square brackets in this case.

**Reply by northern-mindset** [reply]
Then I'm wondering why the square brackets are used in the README docs?

**Reply by HugoDF** [reply]
> Then I'm wondering why the square brackets are used in the README docs?

They're needed when defining methods I think, but yes you're right for straight property names it's not needed

```
{
  ['@click']() {}
}
```

**Reply by northern-mindset** [reply]
See, in my example above I'm successfully using method name without brackets, just in quotes. And it works. I'm sorry for insisting, just trying to figure out any confidential meaning behind this :)

**Reply by HugoDF** [reply]
If it works it works 🙂

I remember we had weird issue when initially building x-spread

**Comment by akaspick** [comment]
I don't see how you would use `$dispatch` via x-spread from the above answer.

I'm attempting to return string literals, but it doesn't seem to work when using `x-spread="checbox"`.  I assume the following isn't supported?  Note, the `item` param is from the use of an outer `x-for` directive.

```
  checkbox: {
      ['@ajax:before']() { return 'startRequest(item)' },
      ['@ajax:success']() { return 'handleSuccess(item, $event, $dispatch)' },
      ['@ajax:error']() { return 'handleError(item, $event)' }
    }
```

Any way to accomplish this with x-spread?

**Reply by HugoDF** [reply]
Based on the example Simone provided you could do:

```js
function Example(){
  return {
    button: {
      ['@click']: 'foo($event, $dispatch)' 
    },
    foo($event, $dispatch) {
    } 
  }
}

**Reply by akaspick** [reply]
That indeed does work.  I was looking at the documentation for `x-spread` that uses `['x-for']() { return 'item in items' }` and mixed up what was written there with what was originally provided in this discussion.  Thanks for pointing me in the right direction.
